Begun adding IRule interface.

hmarp opened the pull request.

hmarp committed:
Begun adding IRule interface.

IMarples commented on PasswordValidator/Rules/IRule.cs:
```diff
@@ -0,0 +1,10 @@
+{
+    public interface IRule
+    {
+        public RuleType ruleType { get; set; }
```
This is missing from all your 'concrete' classes

IMarples commented on PasswordValidator/Validators/AdvancedValidator.cs:
```diff
@@ -1,10 +1,17 @@
     public class AdvancedValidator : IPasswordValidator
     {
         public ValidatorType ValidatorType => ValidatorType.Advanced;
+        public IEnumerable<IRule> Rules { get; set; }
```
private readonly - means it can only be created in the constructor and not assigned to again

hmarp committed:
Injection IEnumerable of rules into Validators.
Added RuleType as field in each IRule concrete class.
Add handling in validator constructors to add selected rules to class field. Modified Validate methods to utilise new Rule format.
Modified unittests to work with new validators dependency of IRule collection.

IMarples commented on PasswordValidator/Program.cs:
```diff
@@ -9,6 +10,11 @@
 builder.Services.AddScoped<IValidatorFactory, PasswordValidatorFactory>();
 builder.Services.AddScoped<IPasswordValidator, SimpleValidator>();
 builder.Services.AddScoped<IPasswordValidator, AdvancedValidator>();
+builder.Services.AddScoped<IRule, AdvancedLengthRule>();
```
Put a whitespace - blank line in between 12 & 13 :-)

IMarples commented on PasswordValidator/Validators/AdvancedValidator.cs:
```diff
@@ -1,18 +1,33 @@
+
+        public AdvancedValidator(IEnumerable<IRule> rules)
+        {
+            var lengthRule = rules.Where(rule => rule.Type == RuleType.AdvancedLength).First() ?? throw new ArgumentException($"Rule of Type {RuleType.AdvancedLength.ToString()} missing");
```
Can we be really clever and put these in a rule builder?

IMarples commented on PasswordValidator_UnitTests/AdvancedValidator_UnitTests.cs:
```diff
@@ -10,7 +12,14 @@ internal class AdvancedValidator_UnitTests
         public AdvancedValidator_UnitTests()
         {
-            _validator = new AdvancedValidator();
+            var rules = new List<IRule>
```
All this has changed now - you will use 'mock' rules - it no longer matter what rules you actually put into it - might be a clue here as to where this is going..

IMarples reviewed (changes requested):
You need a unit test for each of your 'concrete' IRule classes - test each rule individually

IMarples commented on PasswordValidator/Validators/AdvancedValidator.cs:
```diff
@@ -1,18 +1,33 @@
-            bool isValidLength = (password.Length >= 10) && (password.Length <= 18);
-            bool containsSpecialCharacter = password.Any(character => !char.IsLetterOrDigit(character));
-            bool containsEnoughNumbers = password.Where(char.IsDigit).Count() >= 2;
+            foreach (var rule in Rules)
```
This bit is the same for both validators - you can refactor it out into a Base Class they both share?

IMarples commented on PasswordValidator/Validators/AdvancedValidator.cs:
```diff
@@ -1,18 +1,33 @@
     public class AdvancedValidator : IPasswordValidator
     {
         public ValidatorType ValidatorType => ValidatorType.Advanced;
+        private readonly IEnumerable<IRule> Rules;
```
private variables at a class level should be prefixed with _ so this becomes **_rules**

hmarp committed:
Attempted to add Base Validator class.
Attempted to make AdvancedValidator inherit from BaseValidator.
Attempted to make AdvancedValidator_UnitTests use mock IRule collection.

IMarples commented on PasswordValidator_UnitTests/AdvancedValidator_UnitTests.cs:
```diff
@@ -1,47 +1,65 @@
+        public void IsValidAdvancedPassword_AllRulesSatisfied_ShouldReturnTrue()
         {
-            var result = _validator.Validate(advancedPassword);
+            _mockLengthRule.Satisfied(Arg.Any<string>()).Returns(true);
```
You might want to check that your mock objects are getting called?

IMarples commented on PasswordValidator_UnitTests/SimpleValidator_UnitTests.cs:
```diff
@@ -10,7 +12,14 @@ internal class SimpleValidator_UnitTests
         public SimpleValidator_UnitTests()
         {
-            _validator = new SimpleValidator();
+            var rules = new List<IRule>
```
Should be mock objects - not just three - try a collection of one, try an empty collection, try null (it is a public class someone could do that)..

hmarp committed:
Modified Validators to inherit from BaseValidator class, which implements the Validate member.

hmarp committed:
Modified Simple/Advanced validator classes to throw ArgumentNullException if rules parameter is null. Fixed issue with using .First() too.
Rewrote Simple/Advanced validator unittests to use mock IRules.

IMarples commented on PasswordValidator_UnitTests/AdvancedValidator_UnitTests.cs:
```diff
@@ -1,47 +1,150 @@
+        [Test]
+        public void Validate_AllRulesSatisfied_ShouldReturnTrue()
+        {
+            _mockAdvancedLengthRule.Satisfied(Arg.Any<string>()).Returns(true);
```
Don't use `Arg.Any<string>` - be clear about the input you are expecting to receive. I'd put "testPassword" in a variable then replace Arg.Any<string> with the variable.

IMarples commented on PasswordValidator_UnitTests/AdvancedValidator_UnitTests.cs:
```diff
@@ -1,47 +1,150 @@
+            var result = _validator.Validate("testPassword");
+
             Assert.IsTrue(result);
+            _mockAdvancedLengthRule.Received().Satisfied(Arg.Any<string>());
```
Again - replace Arg.Any<string> with the input you expect to receive - that way you will know things are 'correct' - for example it could pass "abc" in not "testPassword"

IMarples replied:
Put an empty line in between 58 & 59

IMarples commented on PasswordValidator_UnitTests/AdvancedValidator_UnitTests.cs:
```diff
@@ -1,47 +1,150 @@
+            _mockAdvancedLengthRule.Received().Satisfied(Arg.Any<string>());
+            _mockAdvancedNumberRule.Received().Satisfied(Arg.Any<string>());
+            _mockCharacterRule.Received().Satisfied(Arg.Any<string>());
+            _mockSimpleNumberRule.DidNotReceive().Satisfied(Arg.Any<string>());
```
Correct use of Arg.Any<string> - only use on `DidNotReceive` as you are not expecting anything.

IMarples commented on PasswordValidator_UnitTests/AdvancedValidator_UnitTests.cs:
```diff
@@ -1,47 +1,150 @@
+
+            var result = _validator.Validate("testPassword");
+
+            Assert.IsFalse(result);
```
Empty line between 78 & 79

IMarples commented on PasswordValidator_UnitTests/AdvancedValidator_UnitTests.cs:
```diff
@@ -1,47 +1,150 @@
+
+            var result = _validator.Validate("testPassword");
+
+            Assert.IsFalse(result);
```
Add empty line

hmarp committed:
Added null input check to each IRule concrete class.
Added unittests for each individual IRule concrete class.

IMarples commented on PasswordValidator_UnitTests/AdvancedValidator_UnitTests.cs:
```diff
@@ -1,47 +1,150 @@
+            if (containsCharacterRule) mockRules.Add(_mockCharacterRule);
+
+            var ex = Assert.Throws<ArgumentException>(() => new AdvancedValidator(mockRules));
+            Assert.AreEqual($"Rule of Type {exceptionRuleName} missing", ex?.Message);
```
good

IMarples commented on PasswordValidator_UnitTests/AdvancedValidator_UnitTests.cs:
```diff
@@ -1,47 +1,150 @@
+        public void CreateAdvancedValidator_InvalidRulesCollection_ShouldThrowArgumentException(bool containsLengthRule, bool containsNumberRule, bool containsCharacterRule, string exceptionRuleName)
+        {
+            var mockRules = new List<IRule>();
+            if (containsLengthRule) mockRules.Add(_mockAdvancedLengthRule);
```
empty line between 134 & 135

IMarples commented on PasswordValidator_UnitTests/PasswordValidatorFactory_UnitTests.cs:
```diff
@@ -11,7 +13,16 @@ internal class PasswordValidatorFactory_UnitTests
         public PasswordValidatorFactory_UnitTests()
         {
-            IPasswordValidator[] validators = new IPasswordValidator[] { new SimpleValidator(), new AdvancedValidator() };
+            var rules = new List<IRule>
```
Should be mocks..

IMarples commented on PasswordValidator_UnitTests/SimpleValidator_UnitTests.cs:
```diff
@@ -1,47 +1,150 @@
     internal class SimpleValidator_UnitTests
     {
-        private readonly IPasswordValidator _validator;
+        private IPasswordValidator? _validator;
```
empty line between 13 & 14

IMarples commented on PasswordValidator_UnitTests/SimpleValidator_UnitTests.cs:
```diff
@@ -1,47 +1,150 @@
+
+            var result = _validator.Validate("testPassword");
+
             Assert.IsTrue(result);
```
Same comments as previous advanced unit test file

IMarples reviewed (changes requested):
Good attempt - it is going in the right direction - few changes

hmarp committed:
Replaced Arg.Any<string>() with explicit references to test password in unittests.
